design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
 
 ## Design Options
 
-### Option A
+### Adapter-Binder Approach
 
-Detail design option A here.
+Overview:
+
+* `Adapters` are bespoke per `Device` as well as per `Protocol`.
+* We term the `Hardware` as per the manufacturer specification to be `Device` + `Adapter`.
+* There is no separate binding.
+* There is no general concrete `Adapter` for any protocol. 
+* The process loop is kicked off by the `Environment`
+
+#### Adapters
+
+The problem with the current implementation is that the `Adapter` concept is too general and the `Binder` mechanism is too limited (1:1) to make linkup to a `Device` powerful or flexible enough.
+
+The `Adapters` in this design are per-protocol & per-device. So they can encapsulate all the logic necessary to drive the specific device that they wrap. 
+
+```python
+class StreamAdapterLinkum95(StreamAdapter):
+    def __init__(self, linkum95_device):
+        if not isinstance(linkum95, SimulatedLinkamT95):
+            raise ValueError("StreamAdapterLinkum95 only wraps the SimulatedLinkamT95 device") 
+        self._device = linkum95_device
+            
+class EPICSAdapterLinkum95(EPICSAdapter):
+    def __init__(self, linkum95_device):
+        if not isinstance(linkum95, SimulatedLinkamT95):
+            raise ValueError("EPICSAdapterLinkum95 only wraps the SimulatedLinkamT95 device")
+        self._device = linkum95_device                     
+```
+
+Adapters are device owners. In the context of the current `linkum95`, the binary flags which are critical to this device would be moved into the associated `StreamAdapterLinkum95`. The state-machine parts of the `Device` are therefore essentially free of the `Hardware` interface. See next section for new-users and external contributions and how we intend to make this separation optional.
+
+#### The Device-Adapter Types
+
+The system should address the needs of the instrument-data group as a priority. That is to implement robust simulators meeting testable and maintainable internal standards. For this device and adapter separation is vital. In this mode of operation the `Device` represents the state-machine independent of the `Adapter` and the `Adapter` turns the `Device` into the `Hardware-Device` which both behaves and communicates as per the real hardware as specified by the manufacturer. For internally developed devices we would expect to see **three factor testing**:
```

Comment:
I like the separation between `Device` and `Hardware-Device`. This describes very accurately what I meant yesterday when I talked about "the model of a device" (this would be the `Device` here). I agree in that this separation opens up good ways for testing.